File: src/de/saschahlusiak/freebloks/view/effects/EffectSet.java
package de.saschahlusiak.freebloks.view.effects;

import java.util.ArrayList;

import javax.microedition.khronos.opengles.GL10;
import de.saschahlusiak.freebloks.view.BoardRenderer;

public class EffectSet extends ArrayList<AbsStoneEffect> implements Effect {
	private static final long serialVersionUID = 1L;

	@Override
	public boolean isEffected(int x, int y) {
		if (size() > 0)
			return get(0).isEffected(x, y);
		return false;
	}

	@Override
	public void render(GL10 gl, BoardRenderer renderer) {
		if (size() > 0)
			get(0).render(gl, renderer);
	}

	@Override
	public void renderShadow(GL10 gl, BoardRenderer renderer) {
		if (size() > 0)
			get(0).renderShadow(gl, renderer);
	}


	@Override
	public boolean isDone() {
		return size() == 0;
	}

	@Override
	public boolean execute(float elapsed) {
		if (size() > 0) {
			boolean b = get(0).execute(elapsed);
			if (get(0).isDone()) {
				remove(0);
				/* don't flag rendering if effect is done */
				return false;
			}
			return b;
		}
		return false;
	}
}

File: src/main/java/de/saschahlusiak/freebloks/view/effects/EffectSet.kt
package de.saschahlusiak.freebloks.view.effects

import de.saschahlusiak.freebloks.view.BoardRenderer
import java.util.*
import javax.microedition.khronos.opengles.GL11

/**
 * Chain of effects, to be played one after each other.
 */
class EffectSet : ArrayList<AbsShapeEffect>(), Effect {
    override fun isEffected(x: Int, y: Int): Boolean {
        return if (size > 0) get(0).isEffected(x, y) else false
    }

    override fun render(gl: GL11, renderer: BoardRenderer) {
        if (size > 0) get(0).render(gl, renderer)
    }

    override fun renderShadow(gl: GL11, renderer: BoardRenderer) {
        if (size > 0) get(0).renderShadow(gl, renderer)
    }

    override fun isDone(): Boolean {
        return size == 0
    }

    override fun execute(elapsed: Float): Boolean {
        if (size > 0) {
            val b = get(0).execute(elapsed)
            if (get(0).isDone()) {
                removeAt(0)
                /* don't flag rendering if effect is done */
                return false
            }
            return b
        }
        return false
    }
}